tests/unit/structure/GEMPLP_unittest.cc

```diff
@@ -0,0 +1,193 @@
+	EXPECT_EQ(arr_max[1],3);
+
+	subset_inds[0] = 1;
+	mplp->max_in_subdimension(arr_tar, subset_inds, arr_max);
+	EXPECT_EQ(arr_max[0],2);
+	EXPECT_EQ(arr_max[1],3);
+
+	SG_UNREF(mplp);
+}
+
+// Test initialization
+TEST(GEMPLP, initialization)
+{
+	CFGTestData* fg_test_data = new CFGTestData();
+	SG_REF(fg_test_data);
+
+	CFactorGraph* fg = fg_test_data->simple_chain_graph();
+	CGEMPLP* mplp = new CGEMPLP(fg);
+	SG_REF(mplp);
+
+	EXPECT_EQ(mplp->m_all_intersections[0][0],0);
+	EXPECT_EQ(mplp->m_all_intersections[0][1],1);
+	EXPECT_EQ(mplp->m_all_intersections[1][0],0);
+	EXPECT_EQ(mplp->m_all_intersections[2][0],1);
+
+	SG_UNREF(fg_test_data);
+	SG_UNREF(fg);
+	SG_UNREF(mplp);
+}
+
+// Test convert message
+TEST(GEMPLP, convert_energy_to_potential)
+{
+	CFGTestData* fg_test_data = new CFGTestData();
+	SG_REF(fg_test_data);
+
+	CFactorGraph* fg = fg_test_data->simple_chain_graph();
+	CGEMPLP* mplp = new CGEMPLP(fg);
+	SG_REF(mplp);
+
+	CFactor* factor = dynamic_cast<CFactor*>(mplp->m_factors->get_element(0));
+	
+	SGNDArray<float64_t> message = mplp->convert_energy_to_potential(factor);
+	
+	EXPECT_EQ(message.len_array, 4);
+	EXPECT_EQ(message.array[0], -0.0);
+	EXPECT_EQ(message.array[1], -0.2);
+	EXPECT_EQ(message.array[2], -0.3);
+	EXPECT_EQ(message.array[3], -0.0);
+
+	SG_UNREF(fg_test_data);
+	SG_UNREF(factor);
+	SG_UNREF(fg);
+	SG_UNREF(mplp);
+}
+
+// Test inference on simple chain graph
+TEST(GEMPLP, simple_chain)
+{
+	CFGTestData* fg_test_data = new CFGTestData();
+	SG_REF(fg_test_data);
+
+	CFactorGraph* fg_simple = fg_test_data->simple_chain_graph();
+	
+	CMAPInference infer_met(fg_simple, GEMPLP);
+	infer_met.inference();
+
+	CFactorGraphObservation* fg_observ = infer_met.get_structured_outputs();
+	SGVector<int32_t> assignment = fg_observ->get_data();
+	SG_UNREF(fg_observ);
+	
+	EXPECT_EQ(assignment[0],0);
+	EXPECT_EQ(assignment[1],0);
+	EXPECT_NEAR(0.4, infer_met.get_energy(), 1E-10);
+
+	SG_UNREF(fg_simple);
+	SG_UNREF(fg_test_data);
+}
+
+// Test inference on random chain graph
+TEST(GEMPLP, random_chain)
+{
+	CFGTestData* fg_test_data = new CFGTestData();
+	SG_REF(fg_test_data);
+
+	SGVector<int32_t> assignment_expected; // expected assignment
+	float64_t min_energy_expected; // expected minimum energy
+
+	CFactorGraph* fg_random = fg_test_data->random_chain_graph(assignment_expected, min_energy_expected);
+		
+	CMAPInference infer_met(fg_random, GEMPLP);
+	infer_met.inference();
+
+	CFactorGraphObservation* fg_observ = infer_met.get_structured_outputs();
+	SGVector<int32_t> assignment = fg_observ->get_data();
+	SG_UNREF(fg_observ);
+	
+	EXPECT_EQ(assignment.size(), assignment_expected.size());
+
+	for (int32_t i = 0; i < assignment.size(); i++)
+		EXPECT_EQ(assignment[i], assignment_expected[i]);
+
+	EXPECT_NEAR(min_energy_expected, infer_met.get_energy(), 1E-10);
+
+	SG_UNREF(fg_random);
+	SG_UNREF(fg_test_data);
+}
+
+// Test with SOSVM
+/* This test is time consuming, uncomment it if needed
+TEST(GEMPLP, sosvm)
```

Comment:
Can we make the data for the unit test more minimal so that the test is not time consuming? Unit tests should be quite minimal, using very toy and simple data sets.